File: linked_list/linked_list.js

class Node{
    constructor(value){
        this.value = value;
        this.next = null;
    }
}

class SinglyLinkedList{
    constructor(){
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(val){
        let newNode = new Node(val);

        if(!this.head){
            this.head = newNode;
            this.tail = newNode;
        } else{
            this.tail.next = newNode;
            this.tail = newNode
        }

        ++this.length;
        return this;
    }

    pop(){
        if(!this.head) return undefined;

        let removedNode = this.head;
        if(this.head === this.tail){
            this.head = null;
            this.tail = null;
        }else{
            let newTail = removedNode;
            while(removedNode.next){
                newTail = removedNode;
                removedNode = removedNode.next;
            }
            newTail.next = null;
            this.tail = newTail;
        }
        --this.length;
        return removedNode;
    }

    shift(){
        if(!this.head) return undefined;

        let removedNode = this.head;
        if(this.head === this.tail){
            this.head = null;
            this.tail = null;
        } else{
            this.head = removedNode.next;
        }
        --this.length;
        return removedNode;
    }

    unshift(val){
        let newNode = new Node(val);

        if(!this.head){
            this.head = newNode;
            this.tail = newNode;
        } else{
            newNode.next = this.head;
            this.head = newNode;
        }

        ++this.length;
        return this;
    }

    get(index){
        if(index < 0 || index >= this.length) return undefined;
        if(!this.head) return undefined;
        let current = this.head;
        let counter = 0;

        while(counter != index){
            current = current.next;
            ++counter;
        }

        return current;
    }

    set(index,value){
        let oldNode = get(index);

        if(oldNode){
            oldNode.value = value;
            return true;
        }
        return false;
    }

    insert(index,value){
        if(index < 0 || index > this.length) return false;
        if(index === 0) return !!this.unshift(value);
        if(index === this.length) return !!this.push(value);

        let newNode = new Node(value);
        let prevNode = this.get(index - 1);
        newNode.next = prevNode.next;
        prevNode.next = newNode;

        ++this.length;
        return true;
    }

    remove(index){
        if(index < 0 || index >= this.length) return undefined;
        if(index === 0) return this.shift();
        if(index === this.length - 1) return this.pop();

        let prevNode = this.get(index - 1);
        let removedNode = prevNode.next;
        prevNode.next = removedNode.next;

        --this.length;
        return removedNode;
    }

    reverse(){
        let current = this.head;
        this.head = this.tail;
        this.tail = current;

        let newNext = null;
        let newPrev;

        for (let i = 0; i < this.length; i++) {
            newPrev = current.next;
            current.next = newNext;

            newNext = current;
            current = newPrev;
        }

        return this;
    }
}

File: miscellenous/data_structure_implementations/linked_list.py
class Node(object):
    def __init__(self,val):
        self.val = val
        self.next = None

class SinglyLinkedList(object):
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def length(self):
        return self.size

    # O(1)
    def insertAtStart(self,val):
        newNode = Node(val)

        if not self.head:
            self.head = self.tail = newNode
        else:
            newNode.next = self.head
            self.head = newNode
        self.size += 1
        return self

    # O(1)
    def insertAtEnd(self,val):
        newNode = Node(val)

        if not self.tail:
            self.head = self.tail = newNode
        else:
            self.tail.next = newNode
            self.tail = newNode
        self.size += 1
        return self

    # O(n)
    def remove(self):
        if not self.tail: return

        current = self.head
        previous = None

        while current.next:
            previous = current
            current = current.next 
        
        if previous:
            previous.next = current.next
            self.tail = previous
        else:
            self.head = current.next

        current.next = None
        self.size -= 1
        return current.val

    def traverse(self):
        if not self.tail: return []
        nodes = []
        current = self.head
        while current:
            nodes.append(current.val)
            current = current.next 
        return nodes


linkList = SinglyLinkedList()

# linkList.insertAtStart(1)
# linkList.insertAtStart(2)
# linkList.insertAtStart(3)
# linkList.insertAtStart(4)
# linkList.insertAtStart(5)
linkList.remove()
# linkList.insertAtEnd(1)
# linkList.insertAtEnd(2)
# linkList.insertAtEnd(3)
# linkList.insertAtEnd(4)
# linkList.insertAtEnd(5)

print(linkList.length())
        
print(linkList.traverse())
